File: examples/maven-spring-preprocessor/java-example/src/main/java/community/flock/wirespec/example/maven/preprocessor/PreProcessor.java
package community.flock.wirespec.example.maven.preprocessor;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.Map;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class PreProcessor implements Function<String, String> {

    ObjectMapper objectMapper = new ObjectMapper();

    @Override
    public String apply(String s) {
        try {
            var openapi = (ObjectNode) objectMapper.readTree(s);
            var paths = openapi.get("paths");
            var spliterator  = Spliterators.spliteratorUnknownSize(paths.fields(), Spliterator.ORDERED);
            var filteredPaths = StreamSupport.stream(spliterator, false)
                    .filter(it -> it.getKey().equals("/pet"))
                    .map(it -> Map.entry(it.getKey(), patchPathItemObject(it.getValue())))
                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
            openapi.set("paths", objectMapper.valueToTree(filteredPaths));
            return objectMapper.writeValueAsString(openapi);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("Cannot process openapi json file", e);
        }
    }

    private JsonNode patchPathItemObject(JsonNode node) {
        patchOperationObject(node.get("post"));
        patchOperationObject(node.get("put"));
        return node;
    }

    private void patchOperationObject(JsonNode node) {
        if (node != null && node.isObject()) {
            ((ObjectNode) node).put("operationId", node.get("operationId") + "Processed"); // Replace "newOperationId" with the desired value
        }
    }

}

File: examples/maven-preprocessor/kotlin-example/src/wirespec/kotlin/community/flock/wirespec/example/maven/preprocessor/PreProcessor.kt
package community.flock.wirespec.example.maven.preprocessor

import community.flock.kotlinx.openapi.bindings.OpenAPIV3
import community.flock.kotlinx.openapi.bindings.OpenAPIV30Model

class PreProcessor : (String) -> String {
    override fun invoke(input: String) =
        (OpenAPIV3.decodeFromString(input) as OpenAPIV30Model)
            .run { copy(paths = paths
                ?.filterKeys { it.value == "/pet" }
                ?.mapValues { it.value.copy(
                    put = it.value.put?.copy(operationId = it.value.put?.operationId?.plus("Processed")),
                    post = it.value.post?.copy(operationId = it.value.post?.operationId?.plus("Processed")),
                ) }
            ) }
            .let(OpenAPIV3::encodeToString)

    }